hal/src/i2c.rs

```diff
@@ -0,0 +1,412 @@
+pub enum Error {
+    /// Arbitration loss
+    Arbitration,
+    /// Bus error
+    Bus,
+    /// Bus busy
+    Busy,
+    /// Not Acknowledge received
+    Nack,
+    /* Overrun, // slave mode only
+     * Pec, // SMBUS mode only
+     * Timeout, // SMBUS mode only
+     * Alert, // SMBUS mode only */
+}
+
+/// I2C peripheral operating in master mode
+pub struct I2c1<PINS> {
+    i2c: I2C1,
+    pins: PINS,
+}
+
+pub struct I2c2<PINS> {
+    i2c: I2C2,
+    pins: PINS,
+}
+
+pub struct I2c3<PINS> {
+    i2c: I2C3,
+    pins: PINS,
+}
+
+macro_rules! busy_wait {
+    ($i2c:expr, $flag:ident, $variant:ident) => {
+        loop {
+            let isr = $i2c.isr.read();
+            let icr = &$i2c.icr;
+
+            if isr.arlo().is_lost() {
+                icr.write(|w| w.arlocf().clear());
+                return Err(Error::Arbitration);
+            } else if isr.berr().is_error() {
+                icr.write(|w| w.berrcf().clear());
+                return Err(Error::Bus);
+            } else if isr.nackf().is_nack() {
+                while $i2c.isr.read().stopf().is_no_stop() {}
+                icr.write(|w| w.nackcf().clear());
+                icr.write(|w| w.stopcf().clear());
+                return Err(Error::Nack);
+            } else if isr.$flag().$variant() {
+                break;
+            }
+        }
+    };
+}
+
+macro_rules! i2c {
+    ($($I2cX:ident: ($I2CX:ident, $i2cXen:ident, $i2cXrst:ident, $i2cXsel:ident,
+                     $I2cXSda:ident, $I2cXScl:ident, $i2cXsclAf:ident, $i2cXsdaAf:ident),)+) => {
+        $(
+            impl<SCL, SDA> $I2cX<(SCL, SDA)> {
+                /// Enables clock and resets the peripheral
+                fn enable_clock(rcc: &mut RCC) {
+                    rcc.apb1enr1.modify(|_, w| w.$i2cXen().enabled());
+                    rcc.apb1rstr1.modify(|_, w| w.$i2cXrst().reset());
+                    rcc.apb1rstr1.modify(|_, w| w.$i2cXrst().clear_bit());
+                }
+
+                fn clock(rcc: &RCC) -> Hertz {
+                    // NOTE(unsafe) atomic read with no side effects
+                    match rcc.ccipr.read().$i2cXsel().variant().unwrap() {
+                        I2C3SEL_A::HSI16 => Hertz(16_000_000),
+                        I2C3SEL_A::SYSCLK => Hertz(sysclk_hz(rcc)), // TODO move the HAL to embedded-time?
+                        I2C3SEL_A::PCLK => Hertz(pclk1_hz(rcc)), // TODO Correct this (HCLK1 / DIV APB1)
+                    }
+                }
+
+                pub fn new(i2c: $I2CX, mut pins: (SCL, SDA), freq: Hertz, rcc: &mut RCC, pullup: bool) -> Self
+                    where
+                    SCL: crate::gpio::sealed::$I2cXScl,
+                    SDA: crate::gpio::sealed::$I2cXSda,
+                    {
+                        assert!(freq.integer() <= 1_000_000);
+
+                        Self::enable_clock(rcc);
+                        pins.0.$i2cXsclAf(pullup);
+                        pins.1.$i2cXsdaAf(pullup);
```

Comment:
I thought I left a comment here, but I don't see it anymore, maybe github removed it because it was too long :sweat_smile: 

I realized there there were lots of back-to-back critical section entries/exits when setting up pins, so I changed the AF traits to accept a critical section in [#112](https://github.com/newAM/stm32wl-hal/pull/112)

This will look a little more like this after those changes:
```rust
cortex_m::interrupt::free(|cs| unsafe {
    pins.0.set_output_type(cs, OutputType::OpenDrain);
    pins.1.set_output_type(cs, OutputType::OpenDrain);
    if (pull) {
        pins.0.set_output_type(cs, Pull::Up);
        pins.1.set_output_type(cs, Pull::Up);
    }
    pins.0.$i2cXsclAf(cs);
    pins.1.$i2cXsdaAf(cs);
});
```

Reply:
Added the I2C AF's in [#114](https://github.com/newAM/stm32wl-hal/pull/114) since I was in there for the RF AF's and UART AF's.

Sorry for the merge conflicts :sweat: let me know if you want me to resolve them.

Reply:
Got it, that's great! No worries I merged it on my side and already changed the implementation to reflect this change.